File: template/ios/ContextMenu/Theme/ContextMenuTheme.swift
// MARK: - ContextMenuTheme.swift
// Автономная тема для кастомного контекстного меню.
// Полностью независима от ChatTheme — может использоваться с любым View.

import UIKit

// MARK: - ContextMenuTheme

public struct ContextMenuTheme {

    // MARK: Backdrop

    public let backdropColor: UIColor
    public let backdropBlurStyle: UIBlurEffect.Style

    // MARK: Emoji panel

    public let emojiPanelBackground: UIColor
    public let emojiPanelCornerRadius: CGFloat
    public let emojiPanelShadowColor: UIColor
    public let emojiPanelShadowOpacity: Float
    public let emojiPanelShadowRadius: CGFloat
    public let emojiFontSize: CGFloat
    public let emojiItemSize: CGFloat

    // MARK: Action menu

    public let menuBackground: UIColor
    public let menuCornerRadius: CGFloat
    public let menuShadowColor: UIColor
    public let menuShadowOpacity: Float
    public let menuShadowRadius: CGFloat
    public let menuSeparatorColor: UIColor

    public let actionTitleFont: UIFont
    public let actionTitleColor: UIColor
    public let actionDestructiveTitleColor: UIColor
    public let actionIconColor: UIColor
    public let actionDestructiveIconColor: UIColor
    public let actionItemHeight: CGFloat
    public let actionHorizontalPadding: CGFloat
    public let actionHighlightColor: UIColor

    // MARK: Animation

    public let openDuration: TimeInterval
    public let closeDuration: TimeInterval
    public let springDamping: CGFloat
    public let springVelocity: CGFloat

    // MARK: Layout

    public let emojiPanelSpacing: CGFloat   // расстояние между emoji-панелью и пузырём
    public let menuSpacing: CGFloat         // расстояние между пузырём и меню
    public let horizontalPadding: CGFloat   // отступы от краёв экрана
    public let verticalPadding: CGFloat     // отступы от safe area
    public let menuWidth: CGFloat

    public init(
        backdropColor: UIColor,
        backdropBlurStyle: UIBlurEffect.Style,
        emojiPanelBackground: UIColor,
        emojiPanelCornerRadius: CGFloat,
        emojiPanelShadowColor: UIColor,
        emojiPanelShadowOpacity: Float,
        emojiPanelShadowRadius: CGFloat,
        emojiFontSize: CGFloat,
        emojiItemSize: CGFloat,
        menuBackground: UIColor,
        menuCornerRadius: CGFloat,
        menuShadowColor: UIColor,
        menuShadowOpacity: Float,
        menuShadowRadius: CGFloat,
        menuSeparatorColor: UIColor,
        actionTitleFont: UIFont,
        actionTitleColor: UIColor,
        actionDestructiveTitleColor: UIColor,
        actionIconColor: UIColor,
        actionDestructiveIconColor: UIColor,
        actionItemHeight: CGFloat,
        actionHorizontalPadding: CGFloat,
        actionHighlightColor: UIColor,
        openDuration: TimeInterval,
        closeDuration: TimeInterval,
        springDamping: CGFloat,
        springVelocity: CGFloat,
        emojiPanelSpacing: CGFloat,
        menuSpacing: CGFloat,
        horizontalPadding: CGFloat,
        verticalPadding: CGFloat,
        menuWidth: CGFloat
    ) {
        self.backdropColor              = backdropColor
        self.backdropBlurStyle          = backdropBlurStyle
        self.emojiPanelBackground       = emojiPanelBackground
        self.emojiPanelCornerRadius     = emojiPanelCornerRadius
        self.emojiPanelShadowColor      = emojiPanelShadowColor
        self.emojiPanelShadowOpacity    = emojiPanelShadowOpacity
        self.emojiPanelShadowRadius     = emojiPanelShadowRadius
        self.emojiFontSize              = emojiFontSize
        self.emojiItemSize              = emojiItemSize
        self.menuBackground             = menuBackground
        self.menuCornerRadius           = menuCornerRadius
        self.menuShadowColor            = menuShadowColor
        self.menuShadowOpacity          = menuShadowOpacity
        self.menuShadowRadius           = menuShadowRadius
        self.menuSeparatorColor         = menuSeparatorColor
        self.actionTitleFont            = actionTitleFont
        self.actionTitleColor           = actionTitleColor
        self.actionDestructiveTitleColor = actionDestructiveTitleColor
        self.actionIconColor            = actionIconColor
        self.actionDestructiveIconColor = actionDestructiveIconColor
        self.actionItemHeight           = actionItemHeight
        self.actionHorizontalPadding    = actionHorizontalPadding
        self.actionHighlightColor       = actionHighlightColor
        self.openDuration               = openDuration
        self.closeDuration              = closeDuration
        self.springDamping              = springDamping
        self.springVelocity             = springVelocity
        self.emojiPanelSpacing          = emojiPanelSpacing
        self.menuSpacing                = menuSpacing
        self.horizontalPadding          = horizontalPadding
        self.verticalPadding            = verticalPadding
        self.menuWidth                  = menuWidth
    }
}

// MARK: - Built-in themes

public extension ContextMenuTheme {

    static let light = ContextMenuTheme(
        backdropColor:               UIColor.black.withAlphaComponent(0.3),
        backdropBlurStyle:           .systemUltraThinMaterial,
        emojiPanelBackground:        UIColor.systemBackground,
        emojiPanelCornerRadius:      16,
        emojiPanelShadowColor:       .black,
        emojiPanelShadowOpacity:     0.10,
        emojiPanelShadowRadius:      12,
        emojiFontSize:               20,
        emojiItemSize:               32,
        menuBackground:              UIColor.systemBackground,
        menuCornerRadius:            12,
        menuShadowColor:             .black,
        menuShadowOpacity:           0.10,
        menuShadowRadius:            16,
        menuSeparatorColor:          UIColor.separator.withAlphaComponent(0.5),
        actionTitleFont:             .systemFont(ofSize: 15),
        actionTitleColor:            .label,
        actionDestructiveTitleColor: .systemRed,
        actionIconColor:             .label,
        actionDestructiveIconColor:  .systemRed,
        actionItemHeight:            38,
        actionHorizontalPadding:     14,
        actionHighlightColor:        UIColor.systemGray5,
        openDuration:                0.40,
        closeDuration:               0.26,
        springDamping:               0.82,
        springVelocity:              0.5,
        emojiPanelSpacing:           6,
        menuSpacing:                 6,
        horizontalPadding:           12,
        verticalPadding:             10,
        menuWidth:                   220
    )

    static let dark = ContextMenuTheme(
        backdropColor:               UIColor.black.withAlphaComponent(0.5),
        backdropBlurStyle:           .systemUltraThinMaterialDark,
        emojiPanelBackground:        UIColor(white: 0.15, alpha: 1),
        emojiPanelCornerRadius:      16,
        emojiPanelShadowColor:       .black,
        emojiPanelShadowOpacity:     0.35,
        emojiPanelShadowRadius:      16,
        emojiFontSize:               20,
        emojiItemSize:               32,
        menuBackground:              UIColor(white: 0.15, alpha: 1),
        menuCornerRadius:            12,
        menuShadowColor:             .black,
        menuShadowOpacity:           0.45,
        menuShadowRadius:            20,
        menuSeparatorColor:          UIColor.white.withAlphaComponent(0.12),
        actionTitleFont:             .systemFont(ofSize: 15),
        actionTitleColor:            UIColor(white: 0.92, alpha: 1),
        actionDestructiveTitleColor: UIColor(red: 1.0, green: 0.35, blue: 0.35, alpha: 1),
        actionIconColor:             UIColor(white: 0.92, alpha: 1),
        actionDestructiveIconColor:  UIColor(red: 1.0, green: 0.35, blue: 0.35, alpha: 1),
        actionItemHeight:            38,
        actionHorizontalPadding:     14,
        actionHighlightColor:        UIColor(white: 0.25, alpha: 1),
        openDuration:                0.40,
        closeDuration:               0.26,
        springDamping:               0.82,
        springVelocity:              0.5,
        emojiPanelSpacing:           6,
        menuSpacing:                 6,
        horizontalPadding:           12,
        verticalPadding:             10,
        menuWidth:                   220
    )
}

File: template/src/shared/ui/context-menu-view/utils/context-menu-theme.ts
import { Platform } from "react-native";

import { ContextMenuTheme } from "../types";

export interface IContextMenuTheme {
  backdropColor: string;
  backdropBlurType:
    "ultraThinMaterial" | "ultraThinMaterialDark" | "light" | "dark";

  emojiPanelBackground: string;
  emojiPanelCornerRadius: number;
  emojiPanelShadowColor: string;
  emojiPanelShadowOpacity: number;
  emojiPanelShadowRadius: number;
  emojiFontSize: number;
  emojiItemSize: number;

  menuBackground: string;
  menuCornerRadius: number;
  menuSeparatorColor: string;
  actionTitleFontSize: number;
  actionTitleColor: string;
  actionDestructiveTitleColor: string;
  actionIconColor: string;
  actionDestructiveIconColor: string;
  actionItemHeight: number;
  actionHorizontalPadding: number;
  actionHighlightColor: string;

  openDuration: number;
  closeDuration: number;
  springDamping: number;
  springVelocity: number;

  emojiPanelSpacing: number;
  menuSpacing: number;
  horizontalPadding: number;
  verticalPadding: number;
  menuWidth: number;

  snapOpenShift: number;
}

export const CONTEXT_MENU_SNAPSHOT_SHADOW = {
  color: "#000000",
  opacity: 0.12,
  radius: 8,
  offsetY: 3,
} as const;

const lightBlurType = Platform.OS === "ios" ? "ultraThinMaterial" : "light";
const darkBlurType = Platform.OS === "ios" ? "ultraThinMaterialDark" : "dark";

export const CONTEXT_MENU_LIGHT_THEME: IContextMenuTheme = {
  backdropColor: "rgba(0, 0, 0, 0.3)",
  backdropBlurType: lightBlurType,
  emojiPanelBackground: "#FFFFFF",
  emojiPanelCornerRadius: 16,
  emojiPanelShadowColor: "#000000",
  emojiPanelShadowOpacity: 0.1,
  emojiPanelShadowRadius: 12,
  emojiFontSize: 20,
  emojiItemSize: 32,
  menuBackground: "#FFFFFF",
  menuCornerRadius: 12,

  menuSeparatorColor: "rgb(217, 217, 217)",
  actionTitleFontSize: 15,
  actionTitleColor: "#000000",

  actionDestructiveTitleColor: "#FF3B30",

  actionIconColor: "rgb(51, 51, 51)",
  actionDestructiveIconColor: "#FF3B30",
  actionItemHeight: 38,
  actionHorizontalPadding: 14,

  actionHighlightColor: "rgb(237, 237, 237)",
  openDuration: 0.4,
  closeDuration: 0.26,
  springDamping: 0.82,
  springVelocity: 0.5,
  emojiPanelSpacing: 6,
  menuSpacing: 6,
  horizontalPadding: 12,
  verticalPadding: 10,
  menuWidth: 220,
  snapOpenShift: 6,
};

export const CONTEXT_MENU_DARK_THEME: IContextMenuTheme = {
  backdropColor: "rgba(0, 0, 0, 0.5)",
  backdropBlurType: darkBlurType,

  emojiPanelBackground: "rgb(38, 38, 38)",
  emojiPanelCornerRadius: 16,
  emojiPanelShadowColor: "#000000",
  emojiPanelShadowOpacity: 0.35,
  emojiPanelShadowRadius: 16,
  emojiFontSize: 20,
  emojiItemSize: 32,
  menuBackground: "rgb(38, 38, 38)",
  menuCornerRadius: 12,
  menuSeparatorColor: "rgba(255, 255, 255, 0.12)",
  actionTitleFontSize: 15,

  actionTitleColor: "rgb(235, 235, 235)",

  actionDestructiveTitleColor: "rgb(255, 89, 89)",
  actionIconColor: "rgb(235, 235, 235)",
  actionDestructiveIconColor: "rgb(255, 89, 89)",
  actionItemHeight: 38,
  actionHorizontalPadding: 14,

  actionHighlightColor: "rgb(64, 64, 64)",
  openDuration: 0.4,
  closeDuration: 0.26,
  springDamping: 0.82,
  springVelocity: 0.5,
  emojiPanelSpacing: 6,
  menuSpacing: 6,
  horizontalPadding: 12,
  verticalPadding: 10,
  menuWidth: 220,
  snapOpenShift: 6,
};

export const resolveContextMenuTheme = (
  name: ContextMenuTheme | string = "light",
): IContextMenuTheme =>
  name.toLowerCase() === "dark"
    ? CONTEXT_MENU_DARK_THEME
    : CONTEXT_MENU_LIGHT_THEME;
